### PR TITLE
move management of public IPs in OpenStack to dedicated class

### PR DESCRIPTION
**Pull request by kivio**
move code of public IP management to one place.

### CODE REVIEW

File: perfkitbenchmarker/openstack/os_network.py [hunk 1]
```diff
@@ -64,3 +66,40 @@ def AllowPort(self, vm, from_port, to_port=None, protocol='tcp'):
...
+        self.__floating_ip_lock = threading.Lock()
+        self.pool = pool
+
+    def allocate(self):
+        with self.__floating_ip_lock:
+            return self.__nclient.floating_ips.create(pool=self.pool)
+
+    def release(self, floating_ip):
+        f_id = floating_ip.id
+        if self.__nclient.floating_ips.get(f_id).fixed_ip:
```
Comment (skschneider) [inline perfkitbenchmarker/openstack/os_network.py]: I can't find NovaClient.floating_ips. Is there another file in your branch that should be in this PR?

Comment (kivio) [inline perfkitbenchmarker/openstack/os_network.py]: NovaClient is just a wrapper of NovaClient from original python-novaclient library which set up credentials and make connection. When you look to code in openstack/utils.py you have simple redirection for all methods to original client class.

Comment (skschneider) [inline perfkitbenchmarker/openstack/os_network.py]: Thanks, I see now.

File: perfkitbenchmarker/openstack/os_network.py [hunk 2]
```diff
@@ -64,3 +66,40 @@ def AllowPort(self, vm, from_port, to_port=None, protocol='tcp'):
...
+            with self.__floating_ip_lock:
+                if not self.__nclient.floating_ips.get(f_id).fixed_ip:
+                    self.__nclient.floating_ips.delete(floating_ip)
+                    while self.__nclient.floating_ips.findall(id=f_id):
+                        time.sleep(1)
```
Comment (skschneider) [inline perfkitbenchmarker/openstack/os_network.py]: Consider a timeout.

File: perfkitbenchmarker/openstack/os_network.py [hunk 3]
```diff
@@ -64,3 +66,40 @@ def AllowPort(self, vm, from_port, to_port=None, protocol='tcp'):
...
     def DisallowAllPorts(self):
         pass
+
+
+class OpenStackPublicNetwork(object):
+
+    def __init__(self, pool):
+        self.__nclient = utils.NovaClient()
+        self.__floating_ip_lock = threading.Lock()
+        self.pool = pool
```
Comment (skschneider) [inline perfkitbenchmarker/openstack/os_network.py]: Could you add a comment about the lock or make the pool variable name more descriptive to match?

File: perfkitbenchmarker/openstack/os_virtual_machine.py [hunk 4]
```diff
@@ -103,9 +107,13 @@ def _PostCreate(self):
...
             instance = self.client.servers.get(self.id)
             status = instance.status
-        self.floating_ip = self.client.floating_ips.create(
-            pool=FLAGS.openstack_public_network)
+
+        self.floating_ip = self.public_network.get_or_create()
         instance.add_floating_ip(self.floating_ip)
+
+        while not self.public_network.is_attached(self.floating_ip):
+            time.sleep(1)
```
Comment (skschneider) [inline perfkitbenchmarker/openstack/os_virtual_machine.py]: Consider a timeout.

**Comment by skschneider** [comment]
Nothing blocking. Let me know when you want this merged.

**Comment by kivio** [comment]
About timeouts - i known currently this is not good written, but i'm looking for global solution for all checks in this driver. Then i thing this code can be merged, and next step is to open new issue with global purpose.

**Comment by voellm** [comment]
Time to merge?

**Comment by kivio** [comment]
I was changing name to more descriptive. @voellm now can you marge it.

**Comment by skschneider** [comment]
Merged, branch can be deleted.

**Pull request merged**